### PR TITLE
🚧  [WIP][Distributed] on adding DistributedActorStorage "header"

### PR DESCRIPTION
**Pull request by ktoso**
Work in progress to prepare "DistributedActorStorage" where we'll store identity and transport.

We'll then move the properties to be implemented via runtime that get the offsets and get the values out of there.

Thanks so much to John's awesome work on DefaultActorStorage this seems it might be very similar and doable!

### CODE REVIEW

File: include/swift/Runtime/Distributed.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// The runtime interface for distributed.
+//
+//===----------------------------------------------------------------------===//
```
Comment (ktoso) [inline include/swift/Runtime/Distributed.h:15]: Very happy to move stuff into it's own header finally <3

File: lib/SILGen/SILGenConstructor.cpp [hunk 2]
```diff
@@ -776,17 +791,47 @@ void SILGenFunction::emitClassConstructorInitializer(ConstructorDecl *ctor) {
...
+
+      // FIXME: can't figure out how to get right type and pass to the thing?
+      auto transportTy = getLoweredTypeForFunctionArgument(transportParam->getType());
+//      SILType transportTy = getLoweredLoadableType(transportParam->getType());
+//      SILType transportTy = SILType::getPrimitiveAddressType(transportParam->getType()->getCanonicalType());
+//      SILType transportTy = getTypeLowering(transportParam->getType())
+//          .getLoweredType();
+
+//      auto transportArg = F.begin()->createFunctionArgument(transportTy, transportParam); // XXXX
+      ManagedValue transportArg = B.createInputFunctionArgument(transportTy, transportParam);
```
Comment (ktoso) [inline lib/SILGen/SILGenConstructor.cpp:830]: I'm super stuck with creating the parameter to pass to the initialize function here.

In this simple take, we want to take the `transport` param of the ctor, and pass it to the `swift_distributedActor_local_initialize` function.

In actual full impl we'd do the dance to invoke the transport to get the identity here as well, and pass both along. But for now I can't figure out how to pass the transport even.

I end up with the following (and some more debug info):

```
Assertion failed: (Index < Length && "Invalid index!"), function operator[], file ArrayRef.h, line 248.
Process 15318 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #4: 0x0000000100815636 swift-frontend`llvm::ArrayRef<swift::SILParameterInfo>::operator[](this=0x0000000132f43748, Index=2) const at ArrayRef.h:248:7
   245 	    /// @name Operator Overloads
   246 	    /// @{
   247 	    const T &operator[](size_t Index) const {
-> 248 	      assert(Index < Length && "Invalid index!");
   249 	      return Data[Index];
   250 	    }
   251
Target 0: (swift-frontend) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #0: 0x00007ff809a09292 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007ff809a392d7 libsystem_pthread.dylib`pthread_kill + 263
    frame #2: 0x00007ff80998dd10 libsystem_c.dylib`abort + 123
    frame #3: 0x00007ff80998d0be libsystem_c.dylib`__assert_rtn + 314
  * frame #4: 0x0000000100815636 swift-frontend`llvm::ArrayRef<swift::SILParameterInfo>::operator[](this=0x0000000132f43748, Index=2) const at ArrayRef.h:248:7
    frame #5: 0x0000000100654518 swift-frontend`swift::SILFunctionConventions::getSILArgumentConvention(this=0x0000000132f437c8, index=2) const at SILArgument.h:36:18
    frame #6: 0x0000000100653eaf swift-frontend`swift::SILBasicBlock::createFunctionArgument(this=0x000000013707eb50, Ty=SILType @ 0x0000000132f43818, D=0x00000001350ac320, disableEntryBlockVerification=false) at SILBasicBlock.cpp:145:32
    frame #7: 0x000000010136c115 swift-frontend`createInputFunctionArgument(B=0x0000000132f4a3e8, type=SILType @ 0x0000000132f438f8, loc=SILLocation @ 0x0000000132f438e8, decl=0x00000001350ac320) at SILGenBuilder.cpp:425:26
    frame #8: 0x000000010136c029 swift-frontend`swift::Lowering::SILGenBuilder::createInputFunctionArgument(this=0x0000000132f4a3e8, type=SILType @ 0x0000000132f43958, decl=0x00000001350ac320) at SILGenBuilder.cpp:458:10
    frame #9: 0x00000001013897c8 swift-frontend`swift::Lowering::SILGenFunction::emitClassConstructorInitializer(this=0x0000000132f4a390, ctor=0x00000001350ac3d8) at SILGenConstructor.cpp:830:37
    frame #10: 0x00000001012e2ec2 swift-frontend`swift::Lowering::SILGenModule::emitFunctionDefinition(this=0x0000000132f4dec0, constant=SILDeclRef @ 0x0000000132f4dd30, f=0x000000013707ea00) at SILGen.cpp:842:37
    frame #11: 0x00000001012ff2ed swift-frontend`(anonymous namespace)::SILGenModuleRAII::~SILGenModuleRAII(this=0x0000000132f4dec0) at SILGen.cpp:2044:13
    frame #12: 0x00000001012ec205 swift-frontend`(anonymous namespace)::SILGenModuleRAII::~SILGenModuleRAII(this=0x0000000132f4dec0) at SILGen.cpp:2036:23
    frame #13: 0x00000001012ebbd2 swift-frontend`swift::ASTLoweringRequest::evaluate(this=0x0000000132f4e6b8, evaluator=0x0000000135011448, desc=ASTLoweringDescriptor @ 0x0000000132f4e4b0) const at SILGen.cpp:2104:1
    frame #14: 0x00000001014f9442 swift-frontend`std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (this=0x0000000132f4e6b8, evaluator=0x0000000135011448, (null)=std::__1::index_sequence<0UL> @ 0x0000000132f4e4a0), (swift::RequestFlags)9>::callDerived<0ul>(swift::Evaluator&, std::__1::integer_sequence<unsigned long, 0ul>) const at SimpleRequest.h:267:24
    frame #15: 0x00000001014f936b swift-frontend`swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(request=0x0000000132f4e6b8, evaluator=0x0000000135011448) at SimpleRequest.h:290:20
    frame #16: 0x000000010130ff4c swift-frontend`llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(this=0x0000000135011448, request=0x0000000132f4e6b8) at Evaluator.h:360:21
    frame #17: 0x00000001012ec41b swift-frontend`llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::operator(this=0x0000000135011448, request=0x0000000132f4e6b8)<swift::ASTLoweringRequest, (void*)0>(swift::ASTLoweringRequest const&) at Evaluator.h:275:12
    frame #18: 0x00000001012ec31e swift-frontend`swift::performASTLowering(mod=0x00000001350183d0, tc=0x0000000133616f90, options=0x000000013500f3f0) at SILGen.cpp:2111:7
    frame #19: 0x000000010013f3db swift-frontend`performCompileStepsPostSema(Instance=0x000000013500ea00, ReturnValue=0x0000000132f4f07c, observer=0x0000000000000000) at FrontendTool.cpp:762:15
    frame #20: 0x000000010013f343 swift-frontend`performAction(this=0x0000000132f4edb0, Instance=0x000000013500ea00)::$_18::operator()(swift::CompilerInstance&) const at FrontendTool.cpp:1260:18
    frame #21: 0x000000010013f29d swift-frontend`bool llvm::function_ref<bool (swift::CompilerInstance&)>::callback_fn<performAction(callable=5149879728, params=0x000000013500ea00)::$_18>(long, swift::CompilerInstance&) at STLExtras.h:185:12
    frame #22: 0x00000001001350f9 swift-frontend`llvm::function_ref<bool (swift::CompilerInstance&)>::operator(this=0x0000000132f4ed18, params=0x000000013500ea00)(swift::CompilerInstance&) const at STLExtras.h:209:12
    frame #23: 0x00000001001338ee swift-frontend`withSemanticAnalysis(Instance=0x000000013500ea00, observer=0x0000000000000000, cont=function_ref<bool (swift::CompilerInstance &)> @ 0x0000000132f4ed18)>) at FrontendTool.cpp:1129:10
    frame #24: 0x000000010012d0ca swift-frontend`performAction(Instance=0x000000013500ea00, ReturnValue=0x0000000132f4f07c, observer=0x0000000000000000) at FrontendTool.cpp:1256:12
    frame #25: 0x000000010011fee0 swift-frontend`performCompile(Instance=0x000000013500ea00, ReturnValue=0x0000000132f4f07c, observer=0x0000000000000000) at FrontendTool.cpp:1308:19
    frame #26: 0x000000010011ee16 swift-frontend`swift::performFrontend(Args=ArrayRef<const char *> @ 0x0000000132f4f2b0, Argv0="/Users/ktoso/code/swift-project/build/Ninja-DebugAssert/swift-macosx-x86_64/bin/swift-frontend", MainAddr=0x000000010002f720, observer=0x0000000000000000) at FrontendTool.cpp:2215:19
    frame #27: 0x0000000100030c64 swift-frontend`run_driver(ExecName=(Data = "swift-frontend", Length = 14), argv=const llvm::ArrayRef<const char *> @ 0x0000000132f50810) at driver.cpp:210:14
    frame #28: 0x000000010002fdd5 swift-frontend`swift::mainEntry(argc_=18, argv_=0x0000000132f53448) at driver.cpp:402:12
    frame #29: 0x000000010002f532 swift-frontend`main(argc_=18, argv_=0x0000000132f53448) at driver.cpp:20:10
    frame #30: 0x000000012eead4d5 dyld`start + 421
(lldb) frame select 5
frame #5: 0x0000000100654518 swift-frontend`swift::SILFunctionConventions::getSILArgumentConvention(this=0x0000000132f437c8, index=2) const at SILArgument.h:36:18
   33  	    assert(silConv.loweredAddresses);
   34  	    return SILArgumentConvention::Indirect_Out;
   35  	  } else {
-> 36  	    auto param = funcTy->getParameters()[index - getNumIndirectSILResults()];
   37  	    return SILArgumentConvention(param.getConvention());
   38  	  }
   39  	}
(lldb) e funcTy->dump()
(sil_function_type type=@convention(method) (@in ActorTransport, @owned FancyWorker) -> @owned FancyWorker
  (input=protocol_type decl=_Distributed.(file).ActorTransport)
  (input=class_type decl=distributed_actor_basic.(file).FancyWorker@/Users/ktoso/code/swift-project/swift/test/Distributed/distributed_actor_basic.swift:10:19)
  (result=class_type decl=distributed_actor_basic.(file).FancyWorker@/Users/ktoso/code/swift-project/swift/test/Distributed/distributed_actor_basic.swift:10:19)
  (substitution_map generic_signature=<nullptr>)
  (substitution_map generic_signature=<nullptr>))
(lldb) frame select 6
frame #6: 0x0000000100653eaf swift-frontend`swift::SILBasicBlock::createFunctionArgument(this=0x000000013707eb50, Ty=SILType @ 0x0000000132f43818, D=0x00000001350ac320, disableEntryBlockVerification=false) at SILBasicBlock.cpp:145:32
   142 	  const SILFunction *Parent = getParent();
   143 	  auto OwnershipKind = ValueOwnershipKind(
   144 	      *Parent, Ty,
-> 145 	      Parent->getConventions().getSILArgumentConvention(getNumArguments())); /// XXX
   146 	  return new (getModule()) SILFunctionArgument(this, Ty, OwnershipKind, D);
   147 	}
   148
```

Comment (ktoso) [reply lib/SILGen/SILGenConstructor.cpp:830]: So yeah we clearly index out of bounds here:

```
(lldb) frame select 5
frame #5: 0x0000000100654518 swift-frontend`swift::SILFunctionConventions::getSILArgumentConvention(this=0x0000000132f437c8, index=2) const at SILArgument.h:36:18
   33  	    assert(silConv.loweredAddresses);
   34  	    return SILArgumentConvention::Indirect_Out;
   35  	  } else {
-> 36  	    auto param = funcTy->getParameters()[index - getNumIndirectSILResults()];
   37  	    return SILArgumentConvention(param.getConvention());
   38  	  }
   39  	}
(lldb) e getNumIndirectSILResults()
(unsigned int) $0 = 0
(lldb) e index
(unsigned int) $1 = 2
(lldb) e funcTy->getParameters()
(llvm::ArrayRef<swift::SILParameterInfo>) $2 = {
  Data = 0x000000013707b098
  Length = 2
}
```

but I really can't figure out how we get to that, been debugging this for a while now hm...

File: lib/SILGen/SILGenDestructor.cpp [hunk 3]
```diff
@@ -77,11 +77,12 @@ void SILGenFunction::emitDestroyingDestructor(DestructorDecl *dd) {
   }
 
 
-  if (cd->isDistributedActor()) {
-    SILBasicBlock *continueBB = createBasicBlock();
-    injectDistributedActorDestructorLifecycleCall(dd, selfValue, continueBB);
-    B.emitBlock(continueBB);
-  }
+//  // FIXME(distributed): enable this again; Assertion failed: (!hasValidInsertionPoint()), function emitBlock, file SILBuilder.h, line 334.
```
Comment (ktoso) [inline lib/SILGen/SILGenDestructor.cpp:80]: not sure why it suddenly started failing... but I'm disabling everything for now just to get the initialization working first.

File: lib/SILGen/SILGenDestructor.cpp [hunk 4]
```diff
@@ -117,6 +118,11 @@ void SILGenFunction::injectDistributedActorDestructorLifecycleCall(
   assert(cd->isDistributedActor() &&
          "only distributed actors have transport lifecycle hooks in deinit");
 
+  // FIXME(distributed): bring back the id storage and remove this hack...!
+  if (cd->lookupDirect(ctx.Id_id).size() > 0)
+    return;
+  // end of FIXME(distributed): bring back the id storage and remove this hack...!
```
Comment (ktoso) [inline lib/SILGen/SILGenDestructor.cpp:124]: this is hack/removed since we don't have a good way to store the ID yet... I'll need to rework it and make it a class I guess until other work lands :-|

File: lib/SILGen/SILGenDistributed.cpp [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+#include "swift/SIL/SILUndef.h"
+#include "swift/SIL/TypeLowering.h"
+
+using namespace swift;
+using namespace Lowering;
+
+/******************************************************************************/
+/******************** DISTRIBUTED ACTOR CONSTRUCTORS **************************/
+/******************************************************************************/
+
```
Comment (ktoso) [inline lib/SILGen/SILGenDistributed.cpp:39]: TODO: move all SILGen distributed stuff here, more centralized

File: lib/SILGen/SILGenConstructor.cpp [hunk 6]
```diff
@@ -776,17 +791,47 @@ void SILGenFunction::emitClassConstructorInitializer(ConstructorDecl *ctor) {
...
-    }
+  // Initialize the distributed actor identity and transport.
+  if (selfClassDecl->isDistributedActor()) {
+    if (ctor->isDistributedActorLocalInit()) {
+      SILLocation PrologueLoc(selfDecl);
+      PrologueLoc.markAsPrologue();
+
+      // TODO(distributed): actually this has to be emitted into every designated distributed
+      //       actor init, not just the "local init one"; it becomes just a
+      //       special case that replaces the no arg initializer for distributed actors
```
Comment (kavon) [inline lib/SILGen/SILGenConstructor.cpp:805]: I think this work should be in a separate PR, with this PR focused on just providing the implementation of DistributedActorStorage as part of the layout of every distributed actor instance, plus built-ins to access its fields. Calling a different built-in to initialize the DistributedActorStorage of a class is kind of a separate concern, imo.

Comment (ktoso) [reply lib/SILGen/SILGenConstructor.cpp:805]: Yeah I'm happy to split things into many PRs -- right now I wanted to share this to ask for help on what I'm stuck with, so the "messy everything I have" PR is just this

File: include/swift/Runtime/Distributed.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+void swift_distributedActor_remote_initialize(DefaultActor *actor);
+
+/// Create a proxy object that will serve as remote distributed actor instance.
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+OpaqueValue* swift_distributedActor_remote_create(
+  /* +1 */OpaqueValue *identity,
+  /* +1 */OpaqueValue *transport,
+  const Metadata *identityType,
+  const Metadata *transportType
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/Runtime/Distributed.h:47]: Ignore this, I'll do a clean PR with many steps and your code -- this PRs status is for me being able to ask for help on https://github.com/apple/swift/pull/38558#discussion_r679255866

File: include/swift/AST/Builtins.def [hunk 8]
```diff
@@ -851,6 +855,14 @@ BUILTIN_MISC_OPERATION_WITH_SILGEN(BuildOrdinarySerialExecutorRef,
...
                                    "buildDefaultActorExecutorRef", "n", Special)
 
+/// Build an AnyActorTransport from a distributed actor reference.
+BUILTIN_MISC_OPERATION_WITH_SILGEN(BuildDistributedActorIdentity,
+                                   "buildDistributedActorIdentity", "n", Special)
+
+/// Build a ActorTransport value from a distributed actor reference.
+BUILTIN_MISC_OPERATION_WITH_SILGEN(BuildDistributedActorTransport,
+                                   "buildDistributedActorTransport", "n", Special)
+
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/AST/Builtins.def:865]: I initially thought we'll have to build the `Any...` there and I also copied naming from the executor access functions -- I think these won't build anything in the end, so well observed we should rename those to just "get" I think 👍

File: include/swift/ABI/MetadataValues.h [hunk 9]
```diff
@@ -47,6 +47,10 @@ enum {
   /// in a default actor.
   NumWords_DefaultActor = 12,
 
+  /// The number of words (in addition to heap-object, and if local also
+  /// actor header) in a distributed actor.
+  NumWords_DistributedActor = 4, // TODO: no idea how many we need?
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/ABI/MetadataValues.h:52]: Seems so :-)

File: include/swift/Runtime/Distributed.h (hunk 1 above)
Comment (ktoso) [reply include/swift/Runtime/Distributed.h:15]: I mean this state of the PR is a very rough WIP I shared to get help on getting unblocked...

My intent was to get your changes merged or cherry pick them here, I'm not worried about the conflicts to be honest -- happy and easy to resolve them

**Comment by ktoso** [comment]
Closing in favor of another implementation approach
